web/src/reducers/notification.js

```diff
@@ -0,0 +1,42 @@
+import {
+  ADD_NOTIFICATION,
+  CLOSE_NOTIFICATION,
+  PROCESS_QUEUE
+} from '../actions/notification';
+
+const initialState = {
+  open: false,
+  messageInfo: {},
+  queue: []
+};
+
+const reducer = (state = initialState, action) => {
+  switch (action.type) {
+    case ADD_NOTIFICATION:
+      return {
+        ...state,
+        queue: [
+          ...state.queue,
+          {
+            message: action.message,
+            key: new Date().getTime()
+          }
+        ]
+      };
+    case CLOSE_NOTIFICATION:
+      return {
+        ...state,
+        open: false
+      };
+    case PROCESS_QUEUE:
+      return {
+        open: true,
+        messageInfo: state.queue[0],
+        queue: state.queue.slice(1)
+      };
```

Comment:
This threw me for a loop for a second because it looked like we could end up in an exception state - if we get `PROCESS_QUEUE` but the `queue` is already empty.  (Technically that's still possible...) but the action checks the state and doesn't dispatch the message if the queue is empty.

I'm curious if that logic belongs in the action creator or the reducer.  We've got similar logic in other action creators too, but I guess I hadn't noticed the separation until now.  What do you think?

Reply:
right, i added that logic to the action creator (https://github.com/18F/cms-hitech-apd/pull/698/files#diff-a49e614a3472c0ef0f58f5c1c69f6847R12) so that you can only get to `PROCESS_QUEUE` in reducer if the queue has entries, but i think it'd be fine to be a little defensive in the reducer as well so i might add that conditional